File: R/linear_regression.R
library(bench)
setwd("/Users/ancavanciupopescu/Desktop/Classes/CS 32/Final project/Data/")
data <- read.csv("linear_regression_data.csv")[,-1]
y <- data[,1]
X <- as.matrix(data[,2:6])

# Linear regression using lm()
linear_regression_package <- function(X, y) {
  model <- lm(y ~ X + 0)
  return(coef(model))
}

# Linear regression from scratch
linear_regression_scratch <- function(X, y) {
  Xb <- cbind(1, X)
  betas <- solve(t(Xb) %*% Xb) %*% (t(Xb) %*% y)
  return(betas)
}


# Measure execution time and memory
results <- bench::mark(
  package = { betas_package <- linear_regression_package(X, y) },
  scratch = { betas_scratch <- linear_regression_scratch(X, y) },
  iterations = 10,
  check = FALSE
)
cat("Execution time for lm:", results$total_time[1], "\n")
cat("Memory allocation for lm:", results$mem_alloc[1]/1e6, "\n")
cat("Coefficients from package:", betas_package, "\n")
cat("Execution time from scratch:", results$total_time[2], "\n")
cat("Memory allocation from scratch:", results$mem_alloc[2]/1e6, "\n")
cat("Coefficients from scratch:", betas_scratch, "\n")

File: Python/linear_regression.py
from memory_profiler import profile
import numpy as np
import time
from sklearn.linear_model import LinearRegression
import pandas as pd

# Load data
path = "/Users/ancavanciupopescu/Desktop/Classes/CS 32/Final project/Data/linear_regression_data.csv"
data = pd.read_csv(path).iloc[:, 1:]
y = data.iloc[:, 0].values
X = data.iloc[:, 1:].values

# Linear regression from scratch
@profile
def linear_regression_from_scratch(X, y):
    X_b = np.hstack([np.ones((X.shape[0], 1)), X])
    beta_hat = np.linalg.inv(X_b.T @ X_b) @ (X_b.T @ y)
    return beta_hat

# Scikit-Learn implementation
@profile
def linear_regression_scikit_learn(X, y):
    model = LinearRegression()
    model.fit(X, y)
    beta_hat = np.hstack([model.intercept_, model.coef_])
    return beta_hat

# Measure execution and memory
start_time_skl = time.time()
beta_hat_skl = linear_regression_scikit_learn(X, y)
end_time_skl = time.time()
skl_time = end_time_skl - start_time_skl
start_time_scratch = time.time()
beta_hat_scratch = linear_regression_from_scratch(X, y)
end_time_scratch = time.time()
scratch_time = end_time_scratch - start_time_scratch
print("Beta hat coefficients using scikit-learn:", beta_hat_skl)
print("Execution time for scikit-learn:", skl_time, "seconds")
print("Beta hat coefficients from scratch:", beta_hat_scratch)
print("Execution time from scratch:", scratch_time, "seconds")